storage-proofs/src/rational_post.rs

```diff
@@ -0,0 +1,473 @@
+
+    fn setup(sp: &Self::SetupParams) -> Result<Self::PublicParams> {
+        Ok(PublicParams {
+            sector_size: sp.sector_size,
+            challenges_count: sp.challenges_count,
+        })
+    }
+
+    fn prove<'b>(
+        _pub_params: &'b Self::PublicParams,
+        pub_inputs: &'b Self::PublicInputs,
+        priv_inputs: &'b Self::PrivateInputs,
+    ) -> Result<Self::Proof> {
+        assert_eq!(
+            pub_inputs.challenges.len(),
+            pub_inputs.commitments.len(),
+            "mismatched challenges and commitments"
+        );
+        let challenges = pub_inputs.challenges;
+
+        let proofs = challenges
+            .iter()
+            .zip(pub_inputs.commitments.iter())
+            .map(|(challenge, commitment)| {
+                let challenged_sector = challenge.sector;
+                let challenged_leaf = challenge.leaf;
+
+                let tree = priv_inputs.trees[challenged_sector as usize];
+                if commitment != &tree.root() {
+                    return Err(Error::InvalidCommitment);
+                }
+
+                Ok(MerkleProof::new_from_proof(
+                    &tree.gen_proof(challenged_leaf as usize),
+                ))
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        Ok(Proof {
+            inclusion_proofs: proofs,
+        })
+    }
+
+    fn verify(
+        pub_params: &Self::PublicParams,
+        pub_inputs: &Self::PublicInputs,
+        proof: &Self::Proof,
+    ) -> Result<bool> {
+        let challenges = pub_inputs.challenges;
+
+        if challenges.len() != pub_inputs.commitments.len() as usize {
+            return Err(Error::MalformedInput);
+        }
+
+        if challenges.len() != proof.inclusion_proofs.len() {
+            return Err(Error::MalformedInput);
+        }
+
+        // validate each proof
+        for ((merkle_proof, challenge), commitment) in proof
+            .inclusion_proofs
+            .iter()
+            .zip(challenges.iter())
+            .zip(pub_inputs.commitments.iter())
+        {
+            let challenged_leaf = challenge.leaf;
+
+            // validate the commitment
+            if merkle_proof.root() != commitment {
+                return Ok(false);
+            }
+
+            // validate the path length
+            if graph_height(pub_params.sector_size as usize / NODE_SIZE)
+                != merkle_proof.path().len()
+            {
+                return Ok(false);
+            }
+
+            if !merkle_proof.validate(challenged_leaf as usize) {
+                return Ok(false);
+            }
+        }
+
+        Ok(true)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Challenge {
+    pub sector: u64,
+    pub leaf: u64,
+}
+
+/// Rational PoSt specific challenge derivation.
+pub fn derive_challenges(
+    challenge_count: usize,
+    sector_size: u64,
+    sector_count: u64,
+    seed: &[u8],
+    faults: &[u64],
```

Comment:
QUESTION: What does one of these `u64` represent? A sector id?

Reply:
yes